### PR TITLE
Calculate transfer functions on grids smaller than data  

### PR DESCRIPTION
**Pull request by talonchandler**
This PR allows transfer functions to be calculated on smaller grids then applied to larger grids.  

This PR uses a `stretched_multiply` strategy for efficient application of a small OTF to a large dataset. This approach approximates the OTF as a piece-wise constant function, and applies this smaller array to datasets without upsampling. This approach allows us to reduce OTF memory requirements by as much as needed, trading OTF computational expense and OTF memory for accuracy.

To uniformly apply this strategy, I have implemented a new function `filter.apply_filter_bank`, which handles these the prepadding, stretched multiplications, matrix-vector bank multiplications, and unpadding. 

These three reconstructions now use a precomputed inverse filter bank and use the new shared `filter.apply_filter_bank`. 
- [x] isotropic_thick_3d
- [x] isotropic_fluorescent_thick_3d
- [x] inplane_oriented_thick_pol3d_vector

### CODE REVIEW

File: waveorder/filter.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
+import itertools
+import numpy as np
```
Comment (ziw-liu) [inline waveorder/filter.py:2]: unused import

Comment (ziw-liu) [reply waveorder/filter.py:2]: Sidenote: we should add linting for this.

File: waveorder/filter.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            "io_filter_bank and i_input_array must have the same number of spatial dimensions."
+        )
+
+    # Ensure the input dimensions match
+    if io_filter_bank.shape[0] != i_input_array.shape[0]:
+        raise ValueError(
+            "io_filter_bank.shape[0] and i_input_array.shape[0] must be the same."
+        )
+
+    I, O = io_filter_bank.shape[:2]
```
Comment (ziw-liu) [inline waveorder/filter.py:74]: What is `I` and `O`? Is it possible to use lower case readable words for these variables?

File: waveorder/filter.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    padded_output_spectrum = torch.zeros((O,) + spatial_dims, dtype=padded_input_spectrum.dtype)
+    for i_idx in range(I):
+        for o_idx in range(O):
+            padded_output_spectrum[o_idx] += stretched_multiply(
+                io_filter_bank[i_idx, o_idx], padded_input_spectrum[i_idx]
+            )
+
+    # Casts to input_array dtype, which typically ignores imaginary part
+    padded_result = torch.fft.ifftn(padded_output_spectrum, dim=fft_dims).type(
+        i_input_array.dtype
```
Comment (ziw-liu) [inline waveorder/filter.py:104]: What is the expected type of `i_input_array`?

**Comment by ziw-liu** [comment]
> These three reconstructions now use a precomputed inverse filter bank

Should this also apply to `isotropic_thin_3d`?

File: waveorder/filter.py (hunk 1 above)
Comment (talonchandler) [reply waveorder/filter.py:2]: Fixed in #182.

File: waveorder/filter.py (hunk 2 above)
Comment (talonchandler) [reply waveorder/filter.py]: changed `I` -> `num_input_channels` and `O` -> `num_output_channels`. Thanks!

File: waveorder/filter.py (hunk 3 above)
Comment (talonchandler) [reply waveorder/filter.py]: Thanks for flagging. 

I was being more general than needed here...we only need real-valued input and outputs for `i_input_array`, and I've updated the docs to make this clearer.

**Comment by talonchandler** [comment]
ZL> Should this also apply to `isotropic_thin_3d`?

Good question. I have added a TODO for this, but I am going to put it off for now because `isotropic_thin_3d` is undergoing a separate debugging effort and I want to avoid muddying the waters.

File: tests/test_filter.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        [[1, 2, 6, 8], [5, 6, 14, 16], [27, 30, 44, 48], [39, 42, 60, 64]]
+    )
+    assert torch.all(result == expected)
+    assert torch.all(
+        filter.stretched_multiply(large_array, large_array) == large_array**2
+    )
+
+    # Test that output dims are correct
+    rand_array_3x3x3 = torch.rand((3, 3, 3))
+    rand_array_100x100x100 = torch.rand((99, 99, 99))
```
Comment (ziw-liu) [inline tests/test_filter.py:39]: Why is this named 100 when the size is 99?

File: waveorder/filter.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+    # Apply the transfer function in the frequency domain
+    fft_dims = [d for d in range(1, i_input_array.ndim)]
+    padded_input_spectrum = torch.fft.fftn(padded_input_array, dim=fft_dims)
+
+    # Matrix-vector multiplication over f
+    # If this is a bottleneck, consider extending `stretched_multiply` to
+    # a `stretched_matrix_multiply` that uses an call like
+    # torch.einsum('io..., i... -> o...', io_filter_bank, padded_input_spectrum)
+    padded_output_spectrum = torch.zeros(
```
Comment (ziw-liu) [inline waveorder/filter.py:97]: Does this need a `device=padded_input_spectrum.device` to work on the GPU?

File: waveorder/filter.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        (num_output_channels,) + spatial_dims,
+        dtype=padded_input_spectrum.dtype,
+    )
+    for input_channel_idx in range(num_input_channels):
+        for output_channel_idx in range(num_output_channels):
+            padded_output_spectrum[output_channel_idx] += stretched_multiply(
```
Comment (ziw-liu) [inline waveorder/filter.py:103]: Torch has some special accelerated calls for block-wise multiply-add. But maybe speed is not a bottleneck here.

File: tests/test_filter.py (hunk 4 above)
Comment (talonchandler) [reply tests/test_filter.py]: Good catch! Vestige from an earlier design. Fixed.

File: waveorder/filter.py (hunk 5 above)
Comment (talonchandler) [reply waveorder/filter.py:97]: Good point. Haven't tested end-to-end yet, but can't hurt. Fixed.

File: waveorder/filter.py (hunk 6 above)
Comment (talonchandler) [reply waveorder/filter.py:104]: In my latest tests this is currently not a bottleneck (we make at most 3 input x 4 output = 12 trips through this loop), but it could become a bottleneck in the future. 

Still I'd appreciate a link to the block-wise multiply-adds that you have in mind. I'm seeing this [`bmm`](https://pytorch.org/docs/stable/generated/torch.bmm.html), but that's not quite what I'd need here.

Comment (ziw-liu) [reply waveorder/filter.py:104]: I was thinking of something like [baddmm](https://pytorch.org/docs/stable/generated/torch.baddbmm.html#torch.baddbmm) with reshaping. But I know realize that the 'blocks' here are super small (2x2?) so the memory access pattern is not really 'blocky'. The current approach is similar to pixel-[shuffle](https://pytorch.org/docs/stable/generated/torch.nn.functional.pixel_shuffle.html#torch.nn.functional.pixel_shuffle)/[unshuffle](https://pytorch.org/docs/stable/generated/torch.nn.functional.pixel_unshuffle.html#torch.nn.functional.pixel_unshuffle) for stretched-multiply followed by a point-wise convolution done here.

Comment (ziw-liu) [reply waveorder/filter.py:104]: (And point-wise convolution is really just a different way to spell 'nn.Linear')

Comment (ziw-liu) [reply waveorder/filter.py:104]: Can it be done via pixel-shuffle -> linear -> pixel-unshuffle?

Comment (talonchandler) [reply waveorder/filter.py:107]: Thanks @ziw-liu. After a quick in-person chat we agreed that this is likely premature optimization, but we agreed that there is likely additional optimization available here.  

I've added a comment that points to these functions so that we can optimize further if we find that this is a bottleneck.

**Review by ziw-liu — APPROVED** [review]
